### PR TITLE
Improve rendering text results

### PR DESCRIPTION
**Pull request by Martchus**
The TAP stuff is actually independent but useful to get results locally, eg.:

screenshot_20180502_181219

### CODE REVIEW

**Comment by Martchus** [comment]
Dealing with overflow by limiting to 3 lines and make the boxes expansible:  
screenshot_20180503_134529

The expanded state is the same as the "needle diff view" for the screenshots. So navigation using arrow keys and hash works like for the needle diff here as well.

**Comment by Martchus** [comment]
You can test on http://e212.suse.de/tests/11993.

Please don't ask how I exported the test results from my local machine and transferred it to e212.

**Comment by foursixnine** [comment]
I wonder how it would look in the end - cherry picks -

**Comment by foursixnine** [comment]
@Martchus: http://phobos.suse.de/tests/11069

**Comment by Martchus** [comment]
@foursixnine Test results for phobos.suse.de links have apparently already been cleaned.

BTW: I prevent that for the test on e212 by revoking write access to the relevant testresults dir.

**Comment by Martchus** [comment]
Ok, the failing test fails for a good reason. The change currently also affects softfailures which shouldn't be the case:
screenshot_20180504_132846

But this way we see at least that mixing regular screenshots and text results is no problem.

**Comment by foursixnine** [comment]
@Martchus I ended up killing gru :)... sadly I cannot apply the patch anymore :( But you got the idea :)....

**Comment by Martchus** [comment]
I re-based and resolved the conflict. I also added an UI test. However, the test is quite limited because this change is mainly about the appearance.

In the current state it is at least prevented that soft-failures and wait-serial results are unaffected. Any other text results will be shown using the new rendering. To limit the new rendering to parser results, the JSON file generated by the parser would need to contain some info that it is parser output. Eg. a source field:  
```
[{"needles":[],"text":"LTP-LTP_net_stress.interface_if4-mtu-change_ip.txt","result":"ok","title":"LTP_net_stress.interface_if4-mtu-change_ip","source":"parser"}]
```

For this, @mudler pointed me to https://github.com/os-autoinst/openQA/blob/master/lib/OpenQA/Parser/Format/LTP.pm#L85, but I haven't achieved anything so far.

---

Besides, for each and every piece of text we currently have an extra file. It would make sense to embed this directly into the JSON file if the text is only a few lines long, eg.:  
```
[{"needles":[],"text_data":"text is directly in JSON","result":"ok","title":"LTP_net_stress.interface_if4-mtu-change_ip","source":"parser"}]
```
(Especially the TAP and xunit parsers produces a lot of one-liners.)

**Comment by richiejp** [comment]
This looks confusing, people won't realise some of the tests steps are from a generated file instead of corresponding to commands in the OpenQA test module. Also we have ~3000 tests in some modules which will be doubled to ~6000. I think we need the processed LTP results (i.e. result_array.json) to be displayed separately from the normal OpenQA output which should also allow extra columns to be added in the future. However the filter is good.

**Comment by Martchus** [comment]
> This looks confusing, people won't realise some of the tests steps are from a generated file instead of corresponding to commands in the OpenQA test module.

Yes. I was already thinking about adding a red row after 'shutdown' which says 'VM terminated - the following results are generated from parser'.

> I think we need the processed LTP results (i.e. result_array.json) to be displayed separately from the normal OpenQA output

That was my first approach. The scraped it because @coolo prefers a uniform view. This has the advantage that we need to implement further features like the filter only once.

> which should also allow extra columns to be added in the future

Even if we make it a separate table, allowing customized columns for different parsers means extra work.

---

@mudler I adjusted the parser test to expect the source field:  
```
diff --git a/t/30-test_parser.t b/t/30-test_parser.t
index 662d2549..98b83a31 100644
--- a/t/30-test_parser.t
+++ b/t/30-test_parser.t
@@ -635,7 +635,7 @@ subtest ltp_parse => sub {
 
 sub serialize_test {
     my ($parser_name, $file, $test_function) = @_;
-    diag "Serialization test for $parser_name and $file with test $test_function";
+    #diag "Serialization test for $parser_name and $file with test $test_function";
     {
         no strict 'refs';    ## no critic
         my $test_result_file = path($FindBin::Bin, "data")->child($file);
@@ -653,6 +653,7 @@ sub serialize_test {
         my $deserialized = $parser_name->new->deserialize($wrote_file->slurp);
         ok "$deserialized" ne "$parser", "Different objects";
 
+        is($parser->results()->first()->{source}, 'parser', 'source parameter set') or diag explain $parser->results();
         $test_function->($parser);
         $test_function->($deserialized);
```

But I have no idea how to change your code to make this tiny change happen without duplicating the code for every format.

I see the following options:  
1. Merge this in the current state.
2. I do the required change in the parser code by duplicating code for each format. No complaints about this in the code review when choosing this option.
3. @mudler takes over to do the required change in the parser code.

I would just go for 1. (see my last comment describing the current state).

**Comment by richiejp** [comment]
> Yes. I was already thinking about adding a red row after 'shutdown' which says 'VM terminated - the following results are generated from parser'.

You will still have 6000 results on one page. I'm sure there are other ways to fix that, but tabs seem fine to me.

> That was my first approach. The scraped it because @coolo prefers a uniform view. This has the advantage that we need to implement further features like the filter only once.

We don't want a uniform view similar to the current test output, we don't like that view, we want a table of results like the one @metan-ucw posted some time back or even like your original demo. You can use unified/generalised data structures (i.e models) and algorithms to avoid reimplementing features like filter or search.

> Even if we make it a separate table, allowing customized columns for different parsers means extra work.

I can guarantee you that this is only the tip of the iceberg. So you should be thinking in terms of "how can I write this so that it can be most easily added to or changed". Because any time you save now by taking the easy path will be insignificant compared to the time spent over the next 5 years by users and developers interacting with your code.

Assuming OpenQA used MVC, adding an extra column would mean storing some meta data in the model which defines how the column's data should be displayed. The view would then read that meta data and the actual data to produce the HTML. This is more work now, but when someone asks for column X or Y, you just have to add a few lines of meta data or we could even produce it from the parser using some schema format. So then it wouldn't be hard coded and the user could set it themselves.

However I am not saying you should do the more ambitious solution first, but that there is no technical argument for using a unified view that outweighs the needs of the users. The users primarily being the Kernel & Networking team. We would be quite happy to get a different, but better solution than what we asked for, but this seems like a worse one. I don't know why, from a user perspective, @coolo prefers having it all on one page, but I think it is overloading one display way too much.

**Comment by coolo** [comment]
For the record: I never said, I want something all in one page. I said I want generic solutions, not LTP quick shots.

**Comment by richiejp** [comment]
Neither does the K&N team, @asmorodskyi needs the same feature for at least Wicked testing, then there are also the fstests and a bunch of other stuff. It might be easier to write a specialised version first then generalise it, but that is an implementation detail not a requirement.

So I would suggest having a separate tab with a plain table with sorting, filtering, one test result per row, name column, result column and output/log column.

**Comment by Martchus** [comment]
@coolo Ok, good to know that you're open for having an extra table. Then I misunderstood you in the last meeting.

---

@richiejp I don't think this PR conflicts with creating a separate table at some point. So if this is not sufficient, I can revive my old PR and make it based on @mudler's generic parser output. However, this parser code is really hard to deal with so I'd like to conclude this PR first. After that I can still add a table, but I need to be able to deal with the generic parser first.

> You can use unified/generalised data structures (i.e models) and algorithms to avoid reimplementing features like filter or search.

I was only talking about the view level. Obviously you not only need a filter algorithm, you also need controls in the front-end eg. for the user to enter a search term. But since we use data tables, we actually have already a filter- and even sorting-enabled view so that shouldn't be a problem.

> Assuming OpenQA used MVC, adding an extra column would mean storing some meta data in the model which defines how the column's data should be displayed.

Unfortunately, in openQA the schema information is hard-coded into the view's HTML/JavaScript code (usually the data table). That's why I meant you don't get columns extensible by the parser for free. But still the effort for adding another column isn't very high.

---

@mudler Ok, I think I could now at least adjust the test so it expects the source parameter:
```
diff --git a/t/30-test_parser.t b/t/30-test_parser.t
index 662d2549..9ba58a67 100644
--- a/t/30-test_parser.t
+++ b/t/30-test_parser.t
@@ -653,6 +653,7 @@ sub serialize_test {
         my $deserialized = $parser_name->new->deserialize($wrote_file->slurp);
         ok "$deserialized" ne "$parser", "Different objects";
 
+        is($_->{source}, 'parser', 'Source parameter set for all steps') for @{$parser->results()->first->{details}};
         $test_function->($parser);
         $test_function->($deserialized);
```
Of course that's not everything because the other occurrences where it currently doesn't expect the source parameter must be fixed as well.

With @foursixnine I tried to modify the constructor or the `to_json`/`to_hash` methods of `openQA/lib/OpenQA/Parser/Result.pm` but without any success. At least it didn't work in all cases.

**Comment by asmorodskyi** [comment]
@richiejp FYI - after discussing this with @foursixnine I came to conclusion that Wicked test will not be able to use this feature. There is a huge difference between LTP and Wicked tests :
1. Wicked tests does not generate output results in JUnit/XUnit format or any other format
2. Wicked tests keep their main test flow inside openQA tests , so Wicked tests care about flow happening on openQA side and can not run without openQA
while LTP tests :
1. Produce their own test output and don't care much about openQA logs ( unless openQA stop functioning at all but this is different story :) ) 
2. LTP tests can be executed fully independently without openQA and just using openQA as Yet Another Framework to execute tests 

I think approach discussed here is more about tests which just using openQA as framework for execution.
And path for Wicked tests is to convience @coolo  and @foursixnine to improve usability for openQA console based tests which is totally different discussion ...

**Comment by Martchus** [comment]
> You will still have 6000 results on one page. I'm sure there are other ways to fix that, but tabs seem fine to me.

Increasing the number of results can be a problem, indeed. The page currently loads already very slow. So we really must change the way the parser stores the generic results (see [my previous comment](https://github.com/os-autoinst/openQA/pull/1639#issuecomment-386642063)). But I would do that in another PR. Like I said, I currently don't know how to deal with the generic parser even for a tiny change.

**Comment by foursixnine** [comment]
@richiejp 

- What do you expect to sort on? Alpabetical order of the tests? - I don't see any benefit, when you can filter (Just looking for an use case)
- What you are seeing now for LTP tests, has been there for sometime, I've mentioned already that we should change that to be shown as steps per LTP run. The way is currently show is that: every element of `$.results.[*].test_fqn` is show as a test module, and the information in `$.results.[*].test.log` is used to display the text info to the user. 

In this regard I have hacked and played a bit with @Martchus 's code to see if I can express myself better. Regardless of wether we go for a separate tab or not, this is completely achievable, but a structure which is already there can be used:

The way it would work in a generic way, is to display: 

Test case | log/output (short) | result
------------ | -------------| -------------

And I would add a time and a misc field, so that the misc field could be used for extra information of the test, and the time would be interesting for filesystem tests for example... or for other xUnit or other formats that also have a time axis defined in their standards. 

screenshot-2018-5-7 openqa openqa-tw-20180411-dev-x86_64-build tw 1638-openqa_from_git 64bit-2g test results 2
screenshot-2018-5-7 openqa openqa-tw-20180411-dev-x86_64-build tw 1638-openqa_from_git 64bit-2g test results 3
screenshot-2018-5-7 openqa openqa-tw-20180411-dev-x86_64-build tw 1638-openqa_from_git 64bit-2g test results 4

I guess is we agree, then we can move forward and adapt this to be a proper table: http://phobos.suse.de/tests/11092#step/openposix/10

And make sure that the filter is able to manage those cases (But a table that has datatables would be better in this case).

**Comment by Martchus** [comment]
@foursixnine It doesn't look nice, but I get what you're aiming for: a small table inside the big table

It would be possible to make this look nice. However, maybe the extra table is still the way to go because that way of making it uniform looks a bit forced.

---

Note that I'm currently blocked with this and I'm not available till next Monday anyways. So like I already said: Either merge this in the current state or @mudler takes over to do the required change in the parser code.

Making it a (separate) table is a thing for another PR. But from my point of view performance issues should be addressed first before trying to make it look more fancy.

---

About the failing Travis test: I just rebased and force-pushed to re-trigger the job. I hope this helps. If not, then it is maybe related to the adjusted test fixtures. Maybe I can use the break during the training to look into it.

**Comment by foursixnine** [comment]
> @foursixnine It doesn't look nice, but I get what you're aiming for: a small table inside the big table

Bingo! :) Doesn't need to look nice, the idea is to prove a point xD. Having a new tab with a separate table, indeed is way better. But it has to be a table.

> Making it a (separate) table is a thing for another PR. But from my point of view performance issues should be addressed first before trying to make it look more fancy.

To solve the performance issues, we would need to rewrite how openQA is storing the information for extra tests, this would mean that we'd have to (most likely) store a json file instead of N txt files (This is the idea that I have in mind) and eventually store that into the database. 

> Either merge this in the current state

Let's better come back to the drawing board. I don't know if it's about reviving your old PR, or simply adapting this one, but merging something that is not clearly defined yet, seems a bit too impromptu imo.

**Comment by richiejp** [comment]
@asmorodskyi OK, I'm not sure if I agree with some of the points, but it is off topic.

@foursixnine It is probably true that it does not make sense to sort on test name, maybe on time however because we sometimes need to identify tests which take too long. I'm not sure about the test suite column, we usually don't upload test results with more than one test suite in, but I'm not saying it is a bad idea either.

> Unfortunately, in openQA the schema information is hard-coded into the view's HTML/JavaScript code (usually the data table). That's why I meant you don't get columns extensible by the parser for free. But still the effort for adding another column isn't very high.

This probably means OpenQA's UI infrastructure needs replacing. Possibly the tools team could find a way of incrementally replacing the current infrastructure with a new one that makes features like this easy.

> To solve the performance issues, we would need to rewrite how openQA is storing the information for extra tests, this would mean that we'd have to (most likely) store a json file instead of N txt files (This is the idea that I have in mind) and eventually store that into the database.

Yes. It makes sense to me to pass a JSON file to the browser (as the model) which is used by some JS library/framework to construct the table dynamically. Again I think this quite similar to what @Martchus did originally, although he was using my JSON format directly instead of a normalised model.

**Comment by Martchus** [comment]
Based on @mudler's change this is now limited to results from parser. I tested this locally. Both, @mudler's changes and my latest adjustments work.

screenshot_20180515_113952

I would merge this first and continue implementing a separate table later.

**Comment by coolo** [comment]
fine with me

File: docs/ExternalResults.asciidoc [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+:toc: left
+:toclevels: 6
+:author: openQA developers
+
+== Introduction
+:parser-formats: <<parser-formats,parsers>>
+:api-endpoint: <<webapi-endpoint,Web Api endpoint>>
+
+From time to time, a test developer might want to use openQA to execute a test
+suite from different test harness than openQA, but still use openQA to setup test
```
[the remark this answers is not in the document]
Comment (Martchus) [reply docs/ExternalResults.asciidoc]: This diff isn't part of the PR.

**Pull request merged**